sdk/trace/batch_span_processor.go

```diff
@@ -167,53 +156,96 @@ func WithBlocking() BatchSpanProcessorOption {
 	}
 }
 
-// processQueue removes spans from the `queue` channel until there is
-// no more data.  It calls the exporter in batches of up to
-// MaxExportBatchSize until all the available data have been processed.
-func (bsp *BatchSpanProcessor) processQueue(batch *[]*export.SpanData) {
+// processQueue removes spans from the `queue` channel until processor
+// is shut down. It calls the exporter in batches of up to MaxExportBatchSize
+// waiting up to ScheduledDelayMillis to form a batch.
+func (bsp *BatchSpanProcessor) processQueue() {
+	timer := time.NewTimer(bsp.o.ScheduledDelayMillis)
+	defer timer.Stop()
+
+	batch := make([]*export.SpanData, 0, bsp.o.MaxExportBatchSize)
+
+	exportSpans := func() {
+		if !timer.Stop() {
+			<-timer.C
+		}
+		timer.Reset(bsp.o.ScheduledDelayMillis)
+
+		if len(batch) > 0 {
+			bsp.e.ExportSpans(context.Background(), batch)
+			batch = batch[:0]
+		}
+	}
+
+loop:
 	for {
-		// Read spans until either the buffer fills or the
-		// queue is empty.
-		for ok := true; ok && len(*batch) < bsp.o.MaxExportBatchSize; {
-			select {
-			case sd := <-bsp.queue:
-				if sd != nil && sd.SpanContext.IsSampled() {
-					*batch = append(*batch, sd)
-				}
-			default:
-				ok = false
+		select {
+		case <-bsp.stopCh:
+			break loop
+		case <-timer.C:
+			exportSpans()
+		case sd := <-bsp.queue:
+			batch = append(batch, sd)
+			if len(batch) == bsp.o.MaxExportBatchSize {
+				exportSpans()
 			}
 		}
+	}
 
-		if len(*batch) == 0 {
-			return
+	go func() {
+		bsp.enqueueWait.Wait()
+		close(bsp.queue)
+	}()
+
+	for {
+		if !timer.Stop() {
+			<-timer.C
 		}
+		const waitTimeout = 30 * time.Second
+		timer.Reset(waitTimeout)
+
+		select {
+		case sd := <-bsp.queue:
+			if sd == nil {
+				exportSpans()
+				return
+			}
 
-		// Send one batch, then continue reading until the
-		// buffer is empty.
-		bsp.e.ExportSpans(context.Background(), *batch)
-		*batch = (*batch)[:0]
+			batch = append(batch, sd)
+			if len(batch) == bsp.o.MaxExportBatchSize {
+				exportSpans()
+			}
+		case <-timer.C:
+			log.Println("bsp.enqueueWait timeout")
+			exportSpans()
+			return
+		}
 	}
 }
 
 func (bsp *BatchSpanProcessor) enqueue(sd *export.SpanData) {
+	if !sd.SpanContext.IsSampled() {
+		return
+	}
+
+	bsp.enqueueWait.Add(1)
```

Comment:
I'm afraid that this use of WaitGroup may panic, it's the comment on
https://golang.org/pkg/sync/#WaitGroup.Add
about Adds concurrent with Waits.

In this method we have to handle two cases: the caller closes the channel, in which case a potential panic, or the caller stops reading the channel but doesn't close it, in which case a potential deadlock.

Some good but not great options:
- defer a call to recover and handle the write-on-closed channel condition
- leak a goroutine to read indefinitely from the channel but do not close it.

My preference is for using a `context.Context` to coordinate cancellation and not closing the channel. At the point of selecting on `bsp.queue <- sd` you'll also select on `<- ctx.Done()`. 

The body of `processQueue` will use a stanza like:

```
ctx, cancel := context.WithCancel(context.Background())
defer cancel()

bsp.cancelCtx = ctx
// any return from this function will unblock the enqueue b/c `<-bsp.cancelCtx.Done()` will succeed
```

Reply:
>I'm afraid that this use of WaitGroup may panic, it's the comment on

I believe this is fine - it panics only when Done or Wait is called before Add which can't happen here.

>At the point of selecting on bsp.queue <- sd you'll also select on <- ctx.Done()

That means that items in the queue will be stuck forever unless we also have "a goroutine to read indefinitely from the channel but do not close it". I think `ctx.Done` does not help here in any way.

Reply:
My statement assumed there would not be any blocking calls to enqueue on the channel. A `ctx.Done()` would unblock the caller in a `select` block. It means two different select blocks would be needed, one for the `BlockOnQueueFull` case and one for the non-blocking case.